lib/Feature/FeatureModelParser.cpp

```diff
@@ -286,10 +275,364 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+        break;
+      default:
+        llvm::errs()
+            << "Wrong indentation or unsupported type of configuration option:'"
+            << To << "'\n";
+        return false;
+      }
+      // Extract the name
+      Name =
+          ToStringRef.substr(Pos + 1, ToStringRef.find(' ', Pos + 1) - Pos - 1)
+              .str();
+
+      // Remove the cardinality
+      if (Name.find_first_of('[') != std::string::npos) {
+        Name = ToStringRef
+                   .substr(Pos + 1, ToStringRef.find('[', Pos + 1) - Pos - 1)
+                   .str();
+
+        if (Name.empty()) {
+          // In this case, the name could also be after the cardinality.
+          // According to the examples provided by S.P.L.O.T., this is a valid
+          // format.
+          auto Remainder =
+              ToStringRef.substr(Pos + 1, ToStringRef.size() - Pos - 1);
+          auto Tokens = Remainder.split(' ');
+          if (!Tokens.second.empty()) {
+            Name = Tokens.second.str();
+          }
+        }
+      }
+
+      // Note that we ignore the ID and use the name of the feature
+      // as unique identifier.
+      if (Name.find_first_of('(') != std::string::npos) {
+        Name = Name.substr(0, Name.find('(', 0));
+      }
+
+      // If there is no name, provide an artificial one
+      if (Name.empty()) {
+        OrGroupCounter++;
+        Name = "group_" + std::to_string(OrGroupCounter);
+      }
+
+      // Create the feature
+      if (IsRoot) {
+        FMB.makeFeature<RootFeature>(Name);
+      } else {
+        FMB.makeFeature<BinaryFeature>(Name, Opt);
+      }
+      IndentationToParentMapping[CurrentIndentationLevel] = Name;
+
+      // Add parent from the upper indentation level if there is one
+      if (LastIndentationLevel != -1 &&
+          CurrentIndentationLevel == RootIndentation) {
+        llvm::errs() << "Only one feature can be root and have the same "
+                        "indentation as root.\n";
+        return false;
+      }
+
+      if (LastIndentationLevel != -1) {
+        auto Parent =
+            IndentationToParentMapping.find(CurrentIndentationLevel - 1);
+        assert(Parent != IndentationToParentMapping.end());
+        FMB.addEdge(Parent->second, Name);
+      }
+
+      // Add the or group to the feature model if it is completely parsed
+      auto OrGroup = OrGroupMapping.find(CurrentIndentationLevel);
+      if (OrGroup != OrGroupMapping.end()) {
+        FMB.emplaceRelationship(std::get<1>(OrGroup->second),
+                                std::get<2>(OrGroup->second),
+                                std::get<0>(OrGroup->second));
+        OrGroupMapping.erase(CurrentIndentationLevel);
+      }
+
+      // Remember the new or group parent if there is one
+      if (Cardinalities.has_value()) {
+        Relationship::RelationshipKind GroupKind =
+            Relationship::RelationshipKind::RK_ALTERNATIVE;
+        if (std::get<1>(Cardinalities.value()) == UINT_MAX) {
```

Comment:
```
warning: comparison of integers of different signs: '__tuple_element_t<1UL, tuple<int, int>>' (aka 'int') and 'unsigned int' [-Wsign-compare]
        if (std::get<1>(Cardinalities.value()) == UINT_MAX) {
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~
```

Reply:
UINT_MAX is in this case '-1' if that is what you wanted you can replace UINT_MAX with -1, or cast UINT_MAX explicitly to int.

Reply:
This is intended to work like this, yes. But you are right. This obfuscates the real intention, so I will replace it with '-1'.

Reply:
I have introduced a constant 'WILDCARD' to make it clear that the '-1' is used as a placeholder for wildcards in the cardinalities.